utils/estimated_histogram.hh

```diff
@@ -47,13 +47,359 @@
+        auto c = count();
+
+        if (!c) {
+            return 0; // no data
+        }
+
+        auto pcount = uint64_t(std::floor(c * quantile));
+        uint64_t elements = 0;
+        for (size_t i = 0; i < NumBuckets - 2; i++) {
+            if (buckets[i]) {
+                elements += buckets[i];
+                if (elements >= pcount) {
+                    return get_bucket_lower_limit(i);
+                }
+            }
+        }
+        return Max; // overflowed value is in the requested percentile
+    }
+
+    /*!
+     * \brief returns the mean histogram value (average of bucket offsets, weighted by count)
+     */
+    uint64_t mean() const {
+        auto lastBucket = NumBuckets - 1;
+        uint64_t elements = 0;
+        uint64_t sum = 0;
+        for (size_t i = 0; i < lastBucket; i++) {
+            long bCount = buckets[i];
+            elements += bCount;
+            sum += bCount * get_bucket_lower_limit(i);
+        }
+        return ((double) (sum + elements -1)/ elements);
+    }
+
+    /*!
+     * \brief returns the number of buckets;
+     */
+    size_t size() const {
+        return NumBuckets;
+    }
+
+    /*!
+     * \brief returns the total number of values inserted
+     */
+    uint64_t count() const {
+        uint64_t sum = 0L;
+        for (size_t i = 0; i < NumBuckets; i++) {
+            sum += buckets[i];
+        }
+        return sum;
+    }
+
+    /*!
+     * \brief multiple all the buckets content in the histogram by a constant
+     */
+    histogram_type& operator*=(double v) {
+        for (size_t i = 0; i < NumBuckets; i++) {
+            buckets[i] *= v;
+        }
+        return *this;
+    }
+
+    template<uint64_t A, uint64_t B, std::vector<uint64_t>::size_type C>
+    friend std::ostream& operator<<(std::ostream& out, const pseudo_floatingpoint_histogram<A, B, C>& h);
+
+    sstring name_of_range(size_t index) const;
+};
+
+template<uint64_t Min, uint64_t Max, std::vector<uint64_t>::size_type NumBuckets>
+inline pseudo_floatingpoint_histogram<Min, Max, NumBuckets> base_estimated_histogram_merge(pseudo_floatingpoint_histogram<Min, Max, NumBuckets> a, const pseudo_floatingpoint_histogram<Min, Max, NumBuckets>& b) {
+    return a.merge(b);
+}
+template<uint64_t Min, uint64_t Max, std::vector<uint64_t>::size_type NumBuckets>
+sstring pseudo_floatingpoint_histogram<Min, Max, NumBuckets>::name_of_range(size_t index) const {
+    sstring s;
+    s += "[";
+
+    s += format("{:d}", get_bucket_lower_limit(index) + 1);
+    s += "..";
+    if (index == buckets.size()) {
+        s += "Inf";
+    } else {
+        s += format("{:d}", get_bucket_upper_limit(index));
+    }
+    s += "]";
+    return s;
+}
+
+template<uint64_t A, uint64_t B, std::vector<uint64_t>::size_type C>
+inline std::ostream& operator<<(std::ostream& out, const pseudo_floatingpoint_histogram<A, B, C>& h) {
+    // only print overflow if there is any
+    size_t name_count;
```

Comment:
This too should be in .cc, with the template parameters passed as normal parameters. You can pass _buckets.data() to avoid having to pass the template.